src/util/bip47.rs

```diff
@@ -0,0 +1,819 @@
+// Bitcoin Dev Kit
+// Written in 2022 by Alekos Filini <alekos.filini@gmail.com>
+//
+// Copyright (c) 2020-2022 Bitcoin Dev Kit Developers
+//
+// This file is licensed under the Apache License, Version 2.0 <LICENSE-APACHE
+// or http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your option.
+// You may not use this file except in accordance with one or both of these
+// licenses.
+
+use std::collections::{BTreeMap, HashMap};
+use std::fmt;
+use std::ops::Deref;
+use std::str::FromStr;
+
+use bitcoin::blockdata::script::Instruction;
+use bitcoin::consensus::encode::serialize;
+use bitcoin::hashes::{sha256, sha512, Hmac, HmacEngine};
+use bitcoin::secp256k1::ecdh::SharedSecret;
+use bitcoin::secp256k1::key::{PublicKey, SecretKey};
+use bitcoin::util::base58;
+use bitcoin::util::bip32;
+use bitcoin::util::psbt;
+use bitcoin::{Address, Network, OutPoint, Script, Transaction, TxIn, Txid};
+
+use crate::blockchain::BlockchainFactory;
+use crate::database::{BatchDatabase, MemoryDatabase};
+use crate::descriptor::template::{DescriptorTemplate, DescriptorTemplateOut, P2Pkh};
+use crate::descriptor::{DescriptorError, Legacy};
+use crate::keys::{DerivableKey, DescriptorSecretKey, DescriptorSinglePriv, ExtendedKey};
+use crate::wallet::coin_selection::DefaultCoinSelectionAlgorithm;
+use crate::wallet::tx_builder::{CreateTx, TxBuilder, TxOrdering};
+use crate::wallet::utils::SecpCtx;
+use crate::wallet::{AddressIndex, SyncOptions, Wallet};
+use crate::{Error as WalletError, KeychainKind, LocalUtxo, TransactionDetails};
+
+#[derive(Copy, Clone, PartialEq, Eq, Debug, PartialOrd, Ord, Hash)]
+pub struct PaymentCode {
```

Comment:
Any specific reasoning behind creating a separate implementation here instead of using the [library in `rust-bitcoin`](https://github.com/rust-bitcoin/rust-bip47/)?

Reply:
It didn't really integrate well with descriptors in my opinion. Adding an extra dependency just to get the `PaymentCode` struct and very few other things felt like a bad idea.

Also, we don't support bitmessage and multisig/bloom filters notifications, so that saved me some time and I only implemented the minimal set of BIP47 that we need.